### PR TITLE
Initial QuadProvider API

### PR DESCRIPTION
**Pull request by ah-OOG-ah**
Adds the QuadProvider API and an example, CubeModel.

### CODE REVIEW

File: src/main/java/com/gtnewhorizons/angelica/compat/nd/Quad.java [hunk 1]
```diff
@@ -31,10 +32,14 @@ public class Quad implements ModelQuadView {
 
     private final Vector3f vectorA = new Vector3f(), vectorB = new Vector3f(), vectorC = new Vector3f();
 
-    private boolean hasColor;
+    public boolean hasColor;
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/compat/nd/Quad.java:35]: why change the visibility? There's a getter, and a setter could be added if needed

File: src/main/java/com/gtnewhorizons/angelica/compat/nd/Quad.java [hunk 2]
```diff
@@ -166,6 +170,12 @@ public void setState(int[] rawBuffer, int offset, BlockRenderer.Flags flags, int
         vectorA.cross(vectorB, vectorC);
 
         normal = ModelQuadFacing.fromVector(vectorC);
+        this.face = ModelQuadFacing.toDirection(normal);
+        try {
+            this.cachedFlags = ModelQuadFlags.getQuadFlags(this);
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/compat/nd/Quad.java:175]: Why is this throwing an NPE?

File: src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java [hunk 3]
```diff
@@ -453,6 +453,17 @@ public enum Mixins {
         .addTargetedMod(TargetedMod.VANILLA)
         .addMixinClasses("mcpatcherforge.ctm_cc.MixinTextureMap")),
 
+    NOVIS_OCULIS(new Builder("Non-Tessellator Quad provider")
+        .setSide(Side.CLIENT)
+        .setPhase(Phase.EARLY)
+        .setApplyIf(() -> AngelicaConfig.injectQPRendering)
+        .addTargetedMod(TargetedMod.VANILLA)
+        .addMixinClasses(
+            "novisoculis.MixinBlockStone",
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java:462]: Let's package these mixins under quadprovider or something similar

File: src/main/java/klaxon/klaxon/novisoculis/CubeModel.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
+package klaxon.klaxon.novisoculis;
```
Comment (mitchej123) [inline src/main/java/klaxon/klaxon/novisoculis/CubeModel.java:1]: Let's keep these under the angelica namespace; might also consider anything that needs to be usable externally to have it available under the API now that we're publishing an API package as well.

File: src/main/java/klaxon/klaxon/novisoculis/CubeModel.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+    public CubeModel(boolean[] colorized) {
+
+        System.arraycopy(colorized, 0, this.colorized, 0, 6);
+    }
+
+    @Override
+    public List<Quad> getQuads(IBlockAccess world, BlockPos pos, Block block, int meta, ForgeDirection dir, Random random) {
+
+        final Quad face = new Quad();
```
Comment (mitchej123) [inline src/main/java/klaxon/klaxon/novisoculis/CubeModel.java:167]: Any reason to avoid using the quad pool?  I'm assuming these get allocated and thrown away afterwards?

File: src/main/java/me/jellysquid/mods/sodium/client/model/light/data/LightDataAccess.java [hunk 6]
```diff
@@ -71,7 +71,7 @@ protected long compute(int x, int y, int z) {
 
         // OPTIMIZE: Do not calculate lightmap data if the block is full and opaque.
         // FIX: Calculate lightmap data for light-emitting or emissive blocks, even though they are full and opaque.
-        final int lm = (fo && !em) ? 0 : block.getMixedBrightnessForBlock(world, x, y, z);
+        final int lm = (false) ? 0 : block.getMixedBrightnessForBlock(world, x, y, z);
```
Comment (mitchej123) [inline src/main/java/me/jellysquid/mods/sodium/client/model/light/data/LightDataAccess.java:74]: ?

File: src/main/java/me/jellysquid/mods/sodium/client/render/occlusion/BlockOcclusionCache.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public boolean shouldDrawSide(Block block, int meta, IBlockAccess view, BlockPos pos, ForgeDirection facing) {
+        BlockPos adjPos = this.cpos;
+        adjPos.set(pos.getX() + facing.offsetX, pos.getY() + facing.offsetY, pos.getZ() + facing.offsetZ);
+
+        Block adjState = view.getBlock(adjPos.x, adjPos.y, adjPos.z);
+
+        return !adjState.isOpaqueCube();
+
+        /*if (selfState.skipRendering(adjState, facing)) {
```
Comment (mitchej123) [inline src/main/java/me/jellysquid/mods/sodium/client/render/occlusion/BlockOcclusionCache.java:37]: likely to be used? If not rm

File: src/main/java/me/jellysquid/mods/sodium/client/world/WorldSlice.java [hunk 8]
```diff
@@ -389,15 +389,15 @@ public static int getLocalChunkIndex(int x, int z) {
...
     // close enough and possibly a behavior change between versions. I also don't know why it's rotationally asymmetric
     public float getBrightness(ForgeDirection direction, boolean shaded) {
+
         if (!shaded) {
             return world.provider.hasNoSky ? 0.9f : 1.0f;
         }
         return switch (direction) {
-            case DOWN -> world.provider.hasNoSky ? 0.9f : 0.5f;
-            case UP -> world.provider.hasNoSky ? 0.9f : 1.0f;
+            case DOWN -> 0.5f;
```
Comment (mitchej123) [inline src/main/java/me/jellysquid/mods/sodium/client/world/WorldSlice.java:397]: Why this change?

File: src/main/java/com/gtnewhorizons/angelica/compat/nd/Quad.java (hunk 2 above)
Comment (ah-OOG-ah) [reply src/main/java/com/gtnewhorizons/angelica/compat/nd/Quad.java:175]: If the quad has facing UNASSIGNED, the facing is null and it NPEs. Vintagium just sets it to DOWN in that case (ordinal 0), so I can change it to that

File: src/main/java/klaxon/klaxon/novisoculis/CubeModel.java (hunk 5 above)
Comment (ah-OOG-ah) [reply src/main/java/klaxon/klaxon/novisoculis/CubeModel.java]: I implemented a quad pool, but from what I can tell unless the writeQuad() implementation releases the quads for reuse, it's the same as just normal allocation. Something to be fixed, perhaps...

File: src/main/java/me/jellysquid/mods/sodium/client/model/light/data/LightDataAccess.java (hunk 6 above)
Comment (ah-OOG-ah) [reply src/main/java/me/jellysquid/mods/sodium/client/model/light/data/LightDataAccess.java:74]: Might have to revert that, I forget if this was an actual fix or just something I was testing; either way needs cleaning

File: src/main/java/klaxon/klaxon/novisoculis/CubeModel.java (hunk 5 above)
Comment (mitchej123) [reply src/main/java/klaxon/klaxon/novisoculis/CubeModel.java]: Yah we'd want to use it in conjunction with release quads... Otherwise no point using the pool it'd grow unbounded.  Worth thinking about, as I don't really want too many allocations in rendering

File: src/main/java/me/jellysquid/mods/sodium/client/model/light/data/LightDataAccess.java (hunk 6 above)
Comment (ah-OOG-ah) [reply src/main/java/me/jellysquid/mods/sodium/client/model/light/data/LightDataAccess.java:74]: Yep, it was just for debugging

File: src/main/java/me/jellysquid/mods/sodium/client/world/WorldSlice.java (hunk 8 above)
Comment (ah-OOG-ah) [reply src/main/java/me/jellysquid/mods/sodium/client/world/WorldSlice.java:397]: Matches Vintagium's rendering. Dunno if it's required, but it works and it's a but cleaner.

File: src/main/java/com/gtnewhorizons/angelica/compat/mojang/Axis.java [hunk 9]
```diff
@@ -9,10 +9,9 @@ public enum Axis {
 
     public static Axis fromDirection(ModelQuadFacing dir) {
         return switch (dir) {
-            case DOWN, UP -> Y;
+            case UP, DOWN, UNASSIGNED -> Y;
```
Comment (embeddedt) [inline src/main/java/com/gtnewhorizons/angelica/compat/mojang/Axis.java:12]: Why does UNASSIGNED have an axis? This seems incorrect.

File: src/main/java/com/gtnewhorizons/angelica/mixins/early/angelica/models/MixinBlockLeaves.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+
+@Mixin(BlockLeaves.class)
+public abstract class MixinBlockLeaves extends Block implements QuadProvider {
+
+    // TODO: Use modern model API
+    private static final ThreadLocal<QuadProvider> model = ThreadLocal.withInitial(() -> new CubeModel(true));
+
+    public List<Quad> getQuads(IBlockAccess world, BlockPos pos, Block block, int meta, ForgeDirection dir, Random random, ObjectPooler<Quad> quadPool) {
+        return model.get().getQuads(world, pos, block, meta, dir, random, quadPool);
```
Comment (embeddedt) [inline src/main/java/com/gtnewhorizons/angelica/mixins/early/angelica/models/MixinBlockLeaves.java:24]: Why does this use a thread-local when stone uses a static object?

**Review by embeddedt — CHANGES_REQUESTED** [review]

File: src/main/java/com/gtnewhorizons/angelica/compat/mojang/Axis.java (hunk 9 above)
Comment (ah-OOG-ah) [reply src/main/java/com/gtnewhorizons/angelica/compat/mojang/Axis.java:12]: Avoids an NPE on launch (for some reason, a quad is made without an axis) and mirrors Vintagium behavior of defaulting to DOWN.

File: src/main/java/com/gtnewhorizons/angelica/mixins/early/angelica/models/MixinBlockLeaves.java (hunk 10 above)
Comment (ah-OOG-ah) [reply src/main/java/com/gtnewhorizons/angelica/mixins/early/angelica/models/MixinBlockLeaves.java:24]: They all use threadlocals, stone and crafting tables just use one stored in CubeModel because they're the same. Leaves inits its own because it needs to be colorized. This is also not a final implementation, nor should it be replicated for more blocks; the final impl for most blocks would be models loaded from JSONs like modern with programmatic models as an option.

File: src/main/java/com/gtnewhorizons/angelica/compat/mojang/Axis.java (hunk 9 above)
Comment (makamys) [reply src/main/java/com/gtnewhorizons/angelica/compat/mojang/Axis.java:12]: The unaligned quad probably comes from a tall grass model or something similar. 1.16.5 assigns UP as the default direction to baked quads since UNASSIGNED is not an option in vanilla. Maybe mirroring this in `Quad` would be more accurate.

File: src/main/java/com/gtnewhorizons/angelica/api/QuadProvider.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+import com.gtnewhorizons.angelica.utils.ObjectPooler;
+import java.util.List;
+import java.util.Random;
+import net.minecraft.block.Block;
+import net.minecraft.world.IBlockAccess;
+import net.minecraftforge.common.util.ForgeDirection;
+
+public interface QuadProvider {
+
+    List<Quad> getQuads(IBlockAccess world, BlockPos pos, Block block, int meta, ForgeDirection dir, Random random, ObjectPooler<Quad> quadPool);
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/api/QuadProvider.java:14]: Do these pooled quads ever get released?  If not the pool will just grow unbounded.

Comment (ah-OOG-ah) [reply src/main/java/com/gtnewhorizons/angelica/api/QuadProvider.java:14]: The pool only stores quads waiting to be used - constantly getting quads is the same as just allocating. The case where you get a growing pool is if you constantly release quads into it. I was going to handle that in a later PR, because if I do that I have to make CapturingTessellator use the right pool too, or it'll continuously dump quads into Blockrenderer's pool and make an actual memory leak.

Comment (mitchej123) [reply src/main/java/com/gtnewhorizons/angelica/api/QuadProvider.java:14]: Sure.  But constantly allocating would get released when it goes out of scope.  Here it never will without a release

File: src/main/java/me/jellysquid/mods/sodium/client/model/quad/properties/ModelQuadFacing.java [hunk 12]
```diff
@@ -61,7 +61,7 @@ public static ModelQuadFacing fromVector(Vector3f normal) {
                 return WEST;
             }
         }
-        return UNASSIGNED;
+        return UP; // Default in 1.16.5
```
Comment (makamys) [inline src/main/java/me/jellysquid/mods/sodium/client/model/quad/properties/ModelQuadFacing.java:64]: This breaks face culling of unaligned quads. This needs to be localized more specifically in `nd.Quad`. On 1.16 it's only `BakedQuad#getFace` that defaults to up.

**Review by makamys — CHANGES_REQUESTED** [review]

File: src/main/java/me/jellysquid/mods/sodium/client/render/pipeline/BlockRenderer.java [hunk 13]
```diff
@@ -53,32 +63,49 @@ public boolean renderModel(IBlockAccess world, RenderBlocks renderBlocks, Block
...
-            final CapturingTessellator tess = (CapturingTessellator) TessellatorManager.get();
-            tess.startDrawingQuads();
-            // RenderBlocks adds the subchunk-relative coordinates as the offset, cancel it out here
+        if (block instanceof QuadProvider qBlock) {
 
-            tess.setOffset(pos);
-            renderBlocks.renderBlockByRenderType(block, pos.x, pos.y, pos.z);
-            final List<Quad> quads = TessellatorManager.stopCapturingToPooledQuads();
-            tess.resetOffset();
+            for (ForgeDirection dir : ForgeDirection.VALID_DIRECTIONS) {
```
[the remark this answers is not in the document]
Comment (ah-OOG-ah) [reply src/main/java/me/jellysquid/mods/sodium/client/render/pipeline/BlockRenderer.java:68]: Why not use ForgeDirection.UNKNOWN? just for clarity?

[the remark this answers is not in the document]
Comment (makamys) [reply src/main/java/me/jellysquid/mods/sodium/client/render/pipeline/BlockRenderer.java:68]: Oh, that would also work. For some reason I forgot `UNKNOWN` existed for a moment there.

File: src/main/java/com/gtnewhorizons/angelica/api/QuadProvider.java (hunk 11 above)
Comment (ah-OOG-ah) [reply src/main/java/com/gtnewhorizons/angelica/api/QuadProvider.java:14]: It goes out of scope in renderQuad, no? ObjectPooler only keeps references to free objects in its pool. If you never `releaseInstance()` it the GC picks it up eventually.

File: src/main/java/me/jellysquid/mods/sodium/client/render/occlusion/BlockOcclusionCache.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * @param block The block in the world
+     * @param meta The meta value of the block
+     * @param view The world view for this render context
+     * @param pos The position of the block
+     * @param facing The facing direction of the side to check
+     * @return True if the block side facing {@param dir} is not occluded, otherwise false
+     */
+    public boolean shouldDrawSide(Block block, int meta, IBlockAccess view, BlockPos pos, ForgeDirection facing) {
```
Comment (makamys) [inline src/main/java/me/jellysquid/mods/sodium/client/render/occlusion/BlockOcclusionCache.java:29]: This should always return true for the UNKNOWN side.

File: src/main/java/com/gtnewhorizons/angelica/api/QuadProvider.java (hunk 11 above)
Comment (ah-OOG-ah) [reply src/main/java/com/gtnewhorizons/angelica/api/QuadProvider.java:14]: Added a release in BlockRenderer

**Review by mitchej123 — APPROVED** [review]
